### DISCUSSION TITLE
How to detect key presses modifer keys

### DISCUSSION THREAD

**Discussion by johnt-star** [microsoft/WindowsAppSDK#2685] [Q&A]
Before I converted my WinUI 3 app to WindowsAppSDK 1.1, I would use CoreWindow to detect key presses and accelerator key presses, as follows:

            Window.Current.CoreWindow.KeyUp += CoreWindow_KeyUp;
            Window.Current.CoreWindow.Dispatcher.AcceleratorKeyActivated += Dispatcher_AcceleratorKeyActivated;

I have yet to find the replacement way to do this in WindowsAppSDK, since CoreWindow is not supported.

I would appreciate if someone can point me in the right direction on how to detect key presses at the Window level and not at the control level.

Thank you in advance!

**Comment by castorix** [comment]
You can use [Microsoft.UI.Input.InputKeyboardSource.GetKeyStateForCurrentThread](https://docs.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.input.inputkeyboardsource.getkeystateforcurrentthread?view=windows-app-sdk-1.1)
I use it in hooks (SetWindowsHookEx and WH_KEYBOARD) for global key states and it works correctly (with _HasFlag_)

**Reply by hawkerm** [reply]
I've found using the `InputKeyboardSource` directly is more reliable due to #4017. My challenge was figuring out how to get access to it from my WinUI app, it seems your app will just already have its own contentisland for the thread:

```cs
        var keyboardSource = InputKeyboardSource.GetForIsland(ContentIsland.FindAllForCurrentThread().First());
        keyboardSource.KeyDown += _keyboardSource_KeyDown;
        keyboardSource.KeyUp += _keyboardSource_KeyUp;
```

**Comment by johnt-star** [comment]
Thanks for the response!

I haven't used hooks before but I will certainly give it a try.
